File: DSA/Asmi_Garg/day1_ques2.java
import java.util.*;

public class PairCounter {

    public static int countPairs(List<Integer> arr, int k) {
        Map<Integer, Integer> remainderCount = new HashMap<>();
        for (int num : arr) {
            int remainder = num % k;
            remainderCount.put(remainder, remainderCount.getOrDefault(remainder, 0) + 1);
        }

        int count = 0;

        for (Map.Entry<Integer, Integer> pair : remainderCount.entrySet()) {
            int remainder = pair.getKey();
            int complement = (k - remainder) % k;

            if (remainder == complement) {
                count += pair.getValue() / 2;
            } else {
                if (remainderCount.containsKey(complement)) {
                    count += Math.min(pair.getValue(), remainderCount.get(complement));
                    remainderCount.put(complement, 0); // Ensures pairs are not counted multiple times
                }
            }
        }
        return count;
    }

    public static void main(String[] args) {
        List<Integer> arr = Arrays.asList(1, 3, 2, 6, 1, 2);
        int k = 3;
        System.out.println(countPairs(arr, k));
    }
}

File: DSA/Shreyashi-Manik-Srivastava/Day1_Ques2.c
#include<stdio.h>

int count_pairs(int arr[], int n, int k){
    int count=0;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            if(arr[i]<arr[j] && (arr[i]+arr[j])%k==0){
                count++;
                printf("Valid pair: (%d, %d)\n", i, j);
            }
        }
    }
    printf("No. of valid pairs: %d\n", count);
}

int main(){
    int n;
    printf("Enter size of the array: ");
    scanf("%d", &n);

    int arr[n];
    printf("Enter array elements:\n");
    for(int i=0; i<n; i++){
        scanf("%d", &arr[i]);
    }

    count_pairs(arr, n, 3);

    return 0;
}